src/vos/vos_obj.c

```diff
@@ -418,6 +430,15 @@ vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 	if (oid.id_shard % 3 == 1 && DAOS_FAIL_CHECK(DAOS_DTX_FAIL_IO))
 		return -DER_IO;
 
+	cont = vos_hdl2cont(coh);
+
+	if (vos_obj_flat_kv_supported(cont, oid) && dkey != NULL && akeys != NULL) {
+		D_ERROR("Akey punch is not supported with flat object types: " DF_UOID "\n",
+			DP_UOID(oid));
+
+		return -DER_INVAL;
```

Comment:
I wonder it should return -INVAL, ENONEXIST or even 0, in case it causing unnecessary failure if the user needs to do group akey punch.

Reply:
Sorry, what is the use case for that?

Reply:
I assume some background ULT might need punch all dkey/akey/recxs for consistency check or fixing object purpose.  Anyway my point is this is trying to punch akey on an non-akey object, it probably should return -ENOEXIST or 0. Though on the other hand, the caller might check the object(FLAT) before punch,  so return -EINVAL might make sense as well. hmm.